internal/utils/afterfunc.go

```diff
@@ -0,0 +1,36 @@
+//go:build !go1.21
+// +build !go1.21
+
+package utils
+
+import (
+	"context"
+	"sync/atomic"
+)
+
+func AfterFunc(ctx context.Context, f func()) (stop func() bool) {
+	if ctx.Done() == nil {
+		return func() bool { return false }
+	}
+
+	var run int32 = 0
+	done := make(chan struct{})
+	go func() {
+		select {
+		case <-ctx.Done():
+			if atomic.CompareAndSwapInt32(&run, 0, 1) {
+				close(done)
+				f()
+			}
+		case <-done:
+		}
+	}()
+
+	return func() bool {
+		if atomic.CompareAndSwapInt32(&run, 0, 1) {
+			close(done)
```

Comment:
is it expected that under certain circumstances the `done` channel will be `closed()` twice (lines 22 and 31)?

Reply:
ah. now i see, there's an atomic to guarantee it.